File: Homework/Van_Le/lesson4/MinStack/MinStack.cs
namespace MinStack
{
    public class MinStack
    {
        private int currentMin;
        private Node head, tail;
        private int NodeCount;

        public MinStack()
        {
			currentMin = int.MaxValue;
            head = new Node();
            tail = head;
            NodeCount = 0;
        }

        public void Push(int value)
        {
            NodeCount++;
            if (value < currentMin)
            {
                currentMin = value;
            }
            if (NodeCount == 1)
            {
                tail.val = value;
                tail.currentMin = currentMin;
            }
            else
            {
                tail.next = new Node(value, currentMin, tail);
                tail = tail.next;
            }
        }

        public void Pop()
        {
            NodeCount--;
            if (NodeCount > 0)
            {
                tail = tail.prev;
                currentMin = tail.currentMin;
            }
            else
            {
                currentMin = int.MaxValue;
                tail.currentMin = currentMin;
            }
        }

        public int Top()
        {
            return tail.val;
        }

        public int GetMin()
        {
            return tail.currentMin;
        }
    }
}

File: Homework/Nam_Nguyen/Lesson4/MinStack.java
package Homework.Nam_Nguyen.Lesson4;

public class MinStack {
    public class Pair {
        private int value;
        private int minSoFar;

        Pair(int value, int minSoFar){
            this.value = value;
            this.minSoFar = minSoFar;
        }
    }

    Pair[] stack;
    int capacity = 0;
    public MinStack() {
        // Constraint: At most 3 * 10^4 calls will be made to push, pop, top, and getMin
        stack = new Pair[30000];
    }
    
    public int min(int a, int b) {
        if(a < b)
            return a;
        return b;
    }    
    public void push(int val) {
        int minSoFar;
        if(capacity > 0)
            minSoFar = stack[capacity-1].minSoFar;
        else 
            minSoFar = Integer.MAX_VALUE;
        minSoFar = min(minSoFar, val);
        Pair element = new Pair(val, minSoFar);
        stack[capacity++] = element;
    }
    
    public void pop() {
        capacity--;
    }
    
    public int top() {
        // Constraint: Methods pop, top and getMin operations will always be called on non-empty stacks.
        // so no need to check stack is empty before get top of stack
        return stack[capacity-1].value;
    }
    
    public int getMin() {
        return stack[capacity-1].minSoFar;
    }
}
